mat64/dense.go

```diff
@@ -512,7 +512,13 @@ func (m *Dense) UnmarshalBinary(data []byte) error {
 	m.mat.Stride = int(cols)
 	m.capRows = int(rows)
 	m.capCols = int(cols)
-	m.mat.Data = use(m.mat.Data, m.mat.Rows*m.mat.Cols)
+
+	size := rows * cols
+	if rows < 0 || cols < 0 || int(size) < 0 || size > maxInt {
+		return matrix.ErrShape
+	}
+
+	m.mat.Data = use(m.mat.Data, int(size))
 
 	for i := range m.mat.Data {
 		err = binary.Read(buf, defaultEndian, &m.mat.Data[i])
```

Comment:
shorter to type and easier to document/enforce in the code, plus: consistency?

actually, I might add a little decoder/encoder helper:

```go
func binWrite(w io.Writer, val interface{}) error {
  return binary.Write(w, binary.LittleEndian, val)
}

func binRead(r io.Reader, valptr interface{}) error {
  return binary.Read(r, binary.LittleEndian, valptr)
}
```

what do you think ?

or, even:

```go
type encoder struct {
  w io.Writer
  err error
}

func (enc *encoder) write(val interface{}) {
  if enc.err != nil {
    return
  }
  enc.err = binary.Write(enc.w, binary.LittleEndian, val)
}

type decoder struct {
  r io.Reader
  err error
}

func (dec *decoder) read(valptr interface{}) {
  if dec.err != nil { return }
  dec.err = binary.Read(dec.r, binary.LittleEndian, valptr)
}
```

to streamline error handling.